### PR TITLE
[Spot] Fix retry forever when filemounts fails

### LINKED ISSUE #2503: [Spot] Failed to catch user file mounts error which causes the retry forever
Observed by @infwinston and a user, that if a file mounts error happens, the spot job will not become `FAILED`, but instead retry forever.

To reproduce:
1. `sky spot launch -n test-fm test.yaml`
2. terminate the spot cluster

```
# test.yaml

file_mounts:
  /data:
    source: gs://existing-bucket/non-exist-folder
    mode: copy
```

The spot job will keep retrying forever, as we did not catch the storage error.
To fix the issue, we can catch the storage exceptions in the following place and raise `ProvisionPrechecksError` to avoid retrying:
https://github.com/skypilot-org/skypilot/blob/3213d3a14ecd86c3022f071e671fcf62c5ca48ba/sky/spot/recovery_strategy.py#L331-L336

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #2503 

~~TODO: deal with network glitch during file mounts~~ (We already do the retry for rsync in our underlying calls)

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] reproducible code in #2503
  - [x] the following steps:
    1. `sky spot launch` a job with file mounts exist
    2. rename the folder on the cloud bucket manually
    3. terminate the spot cluster
    4. The spot job should fail normally.
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Review by cblmemo — COMMENTED** [review]
Briefly went through the PR and left 2 nits. Will submit a more detailed review later 🫡

File: sky/backends/backend_utils.py [hunk 1]
```diff
@@ -1402,10 +1404,24 @@ def parallel_data_transfer_to_nodes(
 
     def _sync_node(runner: 'command_runner.SSHCommandRunner') -> None:
         if cmd is not None:
-            rc, stdout, stderr = runner.run(cmd,
-                                            log_path=log_path,
-                                            stream_logs=stream_logs,
-                                            require_outputs=True)
+            retry_cnt = 0
+            backoff = common_utils.Backoff(initial_backoff=5,
+                                           max_backoff_factor=5)
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/backends/backend_utils.py]: Good catch! Removed. Thanks!

File: sky/backends/backend.py [hunk 2]
```diff
@@ -59,7 +61,12 @@ def provision(
     @timeline.event
     @usage_lib.messages.usage.update_runtime('sync_workdir')
     def sync_workdir(self, handle: _ResourceHandleType, workdir: Path) -> None:
-        return self._sync_workdir(handle, workdir)
+        try:
+            return self._sync_workdir(handle, workdir)
+        except Exception as e:  # pylint: disable=broad-except
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/backends/backend.py:66]: I guess it should be fine to catch all the exceptions. Reason: we want to make it easier for upper level callers to figure out which stage of the `launch` failed.

File: sky/spot/recovery_strategy.py [hunk 3]
```diff
@@ -327,9 +327,42 @@ def _launch(self,
...
+                    # network glitches, so we do not need to retry here.
+                    logger.info('User file mounts failed. Not retrying.\n'
+                                '  Detailed exception: '
+                                f'{common_utils.format_exception(e)}')
+                    # We should raise the exception, no matter whether
+                    # raise_on_failure is True or False, because even though the
+                    # user file mounts suceeded in the previous launch, contents
+                    # in the bucket may have changed, and we should not retry
+                    # the launch, if the user file mounts failed.
+                    raise
```
Comment (cblmemo) [inline sky/spot/recovery_strategy.py:360]: Update docstring for this function with new error types?

Comment (cblmemo) [reply sky/spot/recovery_strategy.py:360]: Also, update docstring for `execution.py::launch`?

**Review by cblmemo — APPROVED** [review]
Looks great to me! Left one nit about docstring 😉

File: sky/backends/backend.py [hunk 4]
```diff
@@ -59,7 +61,12 @@ def provision(
...
+            raise exceptions.FileMountError(
+                f'Failed to sync file mounts for {handle.get_cluster_name()!r}:'
```
Comment (cblmemo) [inline sky/backends/backend.py:68]:
```suggestion
                f'Failed to sync workdir for {handle.get_cluster_name()!r}:'
```

**Review by cblmemo — APPROVED** [review]

File: sky/spot/recovery_strategy.py (hunk 3 above)
Comment (Michaelvll) [reply sky/spot/recovery_strategy.py:356]: Good catch! Fixed. Thanks!

**Comment by Michaelvll** [comment]
Thanks for the review @cblmemo! @concretevitamin, this PR may need your look as well, since it changes some logic in the spot jobs' error handling and need to be double-checked to make sure its correctness.

File: sky/backends/backend_utils.py [hunk 5]
```diff
@@ -1402,10 +1404,23 @@ def parallel_data_transfer_to_nodes(
...
+                                                log_path=log_path,
+                                                stream_logs=stream_logs,
+                                                require_outputs=True)
+                if rc == 255:
+                    if retry_cnt >= MAX_DATA_TRANSFER_RETRY:
+                        break
+                    retry_cnt += 1
+                    backoff_seconds = backoff.current_backoff()
+                    logger.warning(
+                        f'{action_message}: connection error on {runner.ip}. '
```
Comment (concretevitamin) [inline sky/backends/backend_utils.py:1421]: Shall we print stderr too?

File: sky/spot/recovery_strategy.py [hunk 6]
```diff
@@ -240,15 +240,16 @@ def _wait_until_job_starts_on_cluster(self) -> Optional[float]:
 
     def _launch(self,
                 max_retry: Optional[int] = 3,
-                raise_on_failure: bool = True) -> Optional[float]:
+                raise_on_max_retry: bool = True) -> Optional[float]:
```
Comment (concretevitamin) [inline sky/spot/recovery_strategy.py:243]: Is replacing this arg with a new arg with different semantics ok?

https://github.com/skypilot-org/skypilot/blob/42768949f583380a0551bc05f82685351ec5bd24/sky/spot/recovery_strategy.py#L450 

seems to still use the old arg which eventually gets passed up to this new arg.

File: sky/spot/recovery_strategy.py [hunk 7]
```diff
@@ -265,6 +266,10 @@ def _launch(self,
                 1. The optimizer cannot find a feasible solution.
                 2. Precheck errors: invalid cluster name, failure in getting
                 cloud user identity, or unsupported feature.
+            exceptions.FileMountError: The will be raised when the underlying
+                `sky.launch` fails due to file mounts errors only. Since we
+                retry the file mounts in the `sky.launch`, this exception will
+                be raised only when the sources of the file mounts do not exist.
```
Comment (concretevitamin) [inline sky/spot/recovery_strategy.py:272]:
```suggestion
                be raised only when the sources of the file mounts do not exist. (This is checked on initial launch but a cloud bucket source may be changed on a job's recovery attempt.)
```

File: sky/spot/recovery_strategy.py [hunk 8]
```diff
@@ -295,9 +300,7 @@ def _launch(self,
                     exceptions.ResourcesMismatchError) as e:
                 logger.error('Failure happened before provisioning. '
                              f'{common_utils.format_exception(e)}')
-                if raise_on_failure:
```
Comment (concretevitamin) [inline sky/spot/recovery_strategy.py:298]: Left a comment on signature on whether this is ok.

File: sky/spot/recovery_strategy.py [hunk 9]
```diff
@@ -321,15 +324,40 @@ def _launch(self,
                     logger.error(
                         'Failure happened before provisioning. Failover '
                         f'reasons: {reasons_str}')
-                    if raise_on_failure:
```
Comment (concretevitamin) [inline sky/spot/recovery_strategy.py:324]: ditto

File: sky/spot/recovery_strategy.py [hunk 10]
```diff
@@ -321,15 +324,40 @@ def _launch(self,
...
                             f'{common_utils.format_exception(e)})')
+            except exceptions.FileMountError as e:
+                # If the file mounts fails, the failure can be due to:
+                # 1. Errors before the job submission, e.g., invalid file mounts
+                #   -- user changed the bucket content on the cloud manually
+                #   during the launch/recovery. We should not retry the launch,
+                #   as the error is not recoverable.
+                # 2. Preemption during the file mounts. We should be able to
+                #   find the cluster not in UP status, and we can retry the
+                #   launch.
```
Comment (concretevitamin) [inline sky/spot/recovery_strategy.py:338]:
```suggestion
                # 2. Preemption during file mounting. We should be able to
                #   find the cluster not in UP status, and we can retry the
                #   launch.
```

File: sky/spot/recovery_strategy.py [hunk 11]
```diff
@@ -321,15 +324,40 @@ def _launch(self,
...
-                    return None
+                    raise exceptions.ProvisionPrechecksError(reasons=reasons)
                 logger.info('Failed to launch the spot cluster with error: '
                             f'{common_utils.format_exception(e)})')
+            except exceptions.FileMountError as e:
+                # If the file mounts fails, the failure can be due to:
+                # 1. Errors before the job submission, e.g., invalid file mounts
+                #   -- user changed the bucket content on the cloud manually
+                #   during the launch/recovery. We should not retry the launch,
+                #   as the error is not recoverable.
```
Comment (concretevitamin) [inline sky/spot/recovery_strategy.py:335]:
```suggestion
                # If file mounting fails, the failure can be due to:
                # 1. Errors before the job submission, e.g., invalid file mounts
                #   -- user changed the bucket content on the cloud manually
                #   during the launch/recovery. We should not retry the launch,
                #   as the error is not recoverable.
```

File: sky/spot/recovery_strategy.py [hunk 12]
```diff
@@ -321,15 +324,40 @@ def _launch(self,
...
+                #
+                # Pull the actual cluster status from the cloud provider to
+                # determine whether the cluster is preempted.
+                time.sleep(5)
+                (cluster_status,
+                 _) = backend_utils.refresh_cluster_status_handle(
+                     self.cluster_name,
+                     force_refresh_statuses=set(status_lib.ClusterStatus))
+                if cluster_status == status_lib.ClusterStatus.UP:
+                    # Case 1: Errors before the job submission. We should fail
```
Comment (concretevitamin) [inline sky/spot/recovery_strategy.py:348]: Q: why is it the case that an UP status means job has never been submitted?

File: sky/spot/recovery_strategy.py [hunk 13]
```diff
@@ -321,15 +324,40 @@ def _launch(self,
...
+                    # Note: the underlying file mounts will retry for the
+                    # network glitches, so we do not need to retry here.
+                    logger.info('User file mounts failed. Not retrying.\n'
+                                '  Detailed exception: '
+                                f'{common_utils.format_exception(e)}')
+                    raise
+                # Case 2: The cluster is preempted during the launch, retry
+                # the launch.
+                logger.info('The cluster is preempted during the launch. '
+                            'Retrying.')
```
Comment (concretevitamin) [inline sky/spot/recovery_strategy.py:359]: Q:

- Shall we print the cluster status too?
- Is it possible to have a non-preempted cluster reach here with status INIT? E.g., file mounting failed after max retries so the status should be INIT?

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll - mostly nits, a few questions about the state transition.

File: sky/execution.py [hunk 14]
```diff
@@ -477,6 +477,10 @@ def launch(
                 2. Non-empty: iff at least 1 exception from either
                 our pre-checks (e.g., cluster name invalid) or a region/zone
                 throwing resource unavailability.
+        exceptions.FileMountError: the workdir / file mounts failed to sync.
+            Since retry is applied in the underlying file mounts, this will only
+            be raised if the file mounts failed to sync after all retries, e.g.,
+            due to permission errors, or source files not found.
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/execution.py]: Added "network issues persisted longer than max retries" at the end.

File: sky/spot/recovery_strategy.py (hunk 8 above)
Comment (Michaelvll) [reply sky/spot/recovery_strategy.py:298]: This should be raised directly, as retry will not help with these issues.

File: sky/spot/recovery_strategy.py [hunk 15]
```diff
@@ -512,7 +546,8 @@ def recover(self) -> float:
                                              region=launched_region)
                 }
                 # Not using self.launch to avoid the retry until up logic.
-                job_submitted_at = self._launch(raise_on_failure=False)
+                job_submitted_at = self._launch(
```
Comment (concretevitamin) [inline sky/spot/recovery_strategy.py:549]: recover() may crash due to the underlying _launch() raising some non-retryable errors like failed prechecks. Previously recover() will not crash but instead receive None = _launch(…) even if non-retryable errors happened.

Is that ok? Have we tested this code path?

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll. Mostly LGTM, a question about testing non-retryable errors during recover().

**Pull request closed without merging**